### ISSUE TITLE
Rate limit our log messages

### ISSUE THREAD

**Issue by dorlaor** [scylladb/scylladb#1513]
In version 1.2.2, using the AMI.
Checking one of our logs post a crash, I see tons of data in it, there are several problems:
- Many times all shards report the same issue (which all of them hit at once)
- They keep on reporting these issues over and over
- The log isn't informative and even developers wouldn't know what it means (#1522)

Example:

27 05:32:22 ip-172-30-0-124 scylla[4027]:  [shard 15] seastar - Exceptional future ignored: semaphore_timed_out (Semaphore timedout)
Jul 27 05:32:22 ip-172-30-0-124 scylla[4027]:  [shard 16] seastar - Exceptional future ignored: semaphore_timed_out (Semaphore timedout)
Jul 27 05:32:22 ip-172-30-0-124 scylla[4027]:  [shard 16] seastar - Exceptional future ignored: semaphore_timed_out (Semaphore timedout)
Jul 27 05:32:22 ip-172-30-0-124 scylla[4027]:  [shard 21] seastar - Exceptional future ignored: semaphore_timed_out (Semaphore timedout)
Jul 27 05:32:22 ip-172-30-0-124 scylla[4027]:  [shard 22] seastar - Exceptional future ignored: semaphore_timed_out (Semaphore timedout)
Jul 27 05:32:22 ip-172-30-0-124 scylla[4027]:  [shard 23] seastar - Exceptional future ignored: semaphore_timed_out (Semaphore timedout)
Jul 27 05:32:22 ip-172-30-0-124 scylla[4027]:  [shard 18] seastar - Exceptional future ignored: semaphore_timed_out (Semaphore timedout)
Jul 27 05:32:22 ip-172-30-0-124 scylla[4027]:  [shard 23] seastar - Exceptional future ignored: semaphore_timed_out (Semaphore timedout)
Jul 27 05:32:22 ip-172-30-0-124 scylla[4027]:  [shard 26] seastar - Exceptional future ignored: semaphore_timed_out (Semaphore timedout)
Jul 27 05:32:22 ip-172-30-0-124 scylla[4027]:  [shard 26] seastar - Exceptional future ignored: semaphore_timed_out (Semaphore timedout)
Jul 27 05:32:22 ip-172-30-0-124 scylla[4027]:  [shard 28] seastar - Exceptional future ignored: semaphore_timed_out (Semaphore timedout)
Jul 27 05:32:22 ip-172-30-0-124 scylla[4027]:  [shard 26] seastar - Exceptional future ignored: semaphore_timed_out (Semaphore timedout)
Jul 27 05:32:22 ip-172-30-0-124 scylla[4027]:  [shard 28] seastar - Exceptional future ignored: semaphore_timed_out (Semaphore timedout)
Jul 27 05:32:22 ip-172-30-0-124 scylla[4027]:  [shard 29] seastar - Exceptional future ignored: semaphore_timed_out (Semaphore timedout)
Jul 27 05:32:22 ip-172-30-0-124 scylla[4027]:  [shard 27] seastar - Exceptional future ignored: semaphore_timed_out (Semaphore timedout)
Jul 27 05:32:22 ip-172-30-0-124 scylla[4027]:  [shard 26] seastar - Exceptional future ignored: semaphore_timed_out (Semaphore timedout)
Jul 27 05:32:22 ip-172-30-0-124 scylla[4027]:  [shard 29] seastar - Exceptional future ignored: semaphore_timed_out (Semaphore timedout)
Jul 27 05:32:22 ip-172-30-0-124 scylla[4027]:  [shard 27] seastar - Exceptional future ignored: semaphore_timed_out (Semaphore timedout)
Jul 27 05:32:22 ip-172-30-0-124 scylla[4027]:  [shard 27] seastar - Exceptional future ignored: semaphore_timed_out (Semaphore timedout)
Jul 27 05:32:22 ip-172-30-0-124 scylla[4027]:  [shard 11] seastar - Exceptional future ignored: semaphore_timed_out (Semaphore timedout)
Jul 27 05:32:22 ip-172-30-0-124 scylla[4027]:  [shard 18] seastar - Exceptional future ignored: semaphore_timed_out (Semaphore timedout)
Jul 27 05:32:22 ip-172-30-0-124 scylla[4027]:  [shard 18] seastar - Exceptional future ignored: semaphore_timed_out (Semaphore timedout)
Jul 27 05:32:22 ip-172-30-0-124 scylla[4027]:  [shard 16] seastar - Exceptional future ignored: semaphore_timed_out (Semaphore timedout)
Jul 27 05:32:22 ip-172-30-0-124 scylla[4027]:  [shard 4] seastar - Exceptional future ignored: semaphore_timed_out (Semaphore timedout)
Jul 27 05:32:22 ip-172-30-0-124 scylla[4027]:  [shard 15] seastar - Exceptional future ignored: semaphore_timed_out (Semaphore timedout)
Jul 27 05:32:22 ip-172-30-0-124 scylla[4027]:  [shard 9] seastar - Exceptional future ignored: semaphore_timed_out (Semaphore timedout)
Jul 27 05:32:22 ip-172-30-0-124 scylla[4027]:  [shard 19] seastar - Exceptional future ignored: semaphore_timed_out (Semaphore timedout)

Tons of (like 1000s of lines):
Jul 27 05:35:25 ip-172-30-0-124 scylla[4027]:  [shard 30] seastar - Exceptional future ignored: std::runtime_error (sstable inactive read queue overloaded)
Jul 27 05:35:25 ip-172-30-0-124 scylla[4027]:  [shard 16] seastar - Exceptional future ignored: std::runtime_error (sstable inactive read queue overloaded)
Jul 27 05:35:25 ip-172-30-0-124 scylla[4027]:  [shard 13] seastar - Exceptional future ignored: std::runtime_error (sstable inactive read queue overloaded)
Jul 27 05:35:25 ip-172-30-0-124 scylla[4027]:  [shard 18] seastar - Exceptional future ignored: std::runtime_error (sstable inactive read queue overloaded)
Jul 27 05:35:25 ip-172-30-0-124 scylla[4027]:  [shard 29] seastar - Exceptional future ignored: std::runtime_error (sstable inactive read queue overloaded)

**Comment by gleb-cloudius** [comment]
On Wed, Jul 27, 2016 at 03:59:41AM -0700, Dor Laor wrote:
> Checking one of our logs post a crash, I see tons of data in it, there are several problems:
>   - Many times all shards report the same issue (which all of them hit at once)
Should we coordinate logging between all shards and make everything slow?

>   - They keep on reporting these issues over and over
That what rate limiter with duplication removal suppose to fix.

>   - The log isn't informative and even developers wouldn't know what it means.
> 
What do you propose? For logger to analyze the problem and point out to
the problematic line in the code like that:

Jul 27 05:35:25 ip-172-30-0-124 scylla[4027]:  the bug is in file buggy_code.cc at line 666?

--
			Gleb.

**Comment by dorlaor** [comment]
On Wed, Jul 27, 2016 at 2:05 PM, Gleb Natapov <notifications@github.com>
wrote:

> On Wed, Jul 27, 2016 at 03:59:41AM -0700, Dor Laor wrote:
> > Checking one of our logs post a crash, I see tons of data in it, there
> are several problems:
> > - Many times all shards report the same issue (which all of them hit at
> once)
> Should we coordinate logging between all shards and make everything slow?
>

I bet you can come up with something which will be fast. If your rate
limiter
will work, it may be less urgent.


>
> > - They keep on reporting these issues over and over
> That what rate limiter with duplication removal suppose to fix.


> > - The log isn't informative and even developers wouldn't know what it
> means.
> >
> What do you propose? For logger to analyze the problem and point out to
> the problematic line in the code like that:
>
> Jul 27 05:35:25 ip-172-30-0-124 scylla[4027]: the bug is in file
> buggy_code.cc at line 666?
>


It still better than 'exceptional future'. Everyone knows what's 'bug' is.
Gleb, why are you arguing. At best you can claim that at the point of this
issue there is no information that can be added and there is no possible way
to add this info which I believe there is such

**Comment by gleb-cloudius** [comment]
On Wed, Jul 27, 2016 at 04:11:57AM -0700, Dor Laor wrote:
> > > - The log isn't informative and even developers wouldn't know what it
> > means.
> > >
> > What do you propose? For logger to analyze the problem and point out to
> > the problematic line in the code like that:
> >
> > Jul 27 05:35:25 ip-172-30-0-124 scylla[4027]: the bug is in file
> > buggy_code.cc at line 666?
> >
> 
> 
> It still better than 'exceptional future'. Everyone knows what's 'bug' is.
Exceptional future is useful for developer since it points to an error
that was not properly handled. If it would be possible to add more info
about the error it would have been great, but at the point it is printed
it is not possible.

> Gleb, why are you arguing. At best you can claim that at the point of this
> issue there is no information that can be added and there is no possible way
> to add this info which I believe there is such
> 
The issue should be formulated in such a way that it can be eventually
closed, what is the criteria for this one to be closed?

--
			Gleb.

**Comment by dorlaor** [comment]
@gleb-cloudius agree. I changed the title to add a rate limit and optioned #1522 above log message.

**Comment by bentsi** [comment]
Had this been already implemented? I see in the DB log messages: 
`Rate-limit: supressed X backtraces on shard 13`

**Comment by denesb** [comment]
Semaphore timeout log messages are now rate-limited.
